Feat cluster variable management

mathias-vandaele opened the pull request:
## Description

# PR: Cluster Variable Management (Create / Update / Delete)

## Overview

This PR introduces end-to-end support for **cluster variable management**, including create, update, and delete operations across multiple components — client, gateway (REST + broker), engine processing/state, exporters, persistence, security, and testing.

---

## Video presentation

https://www.loom.com/share/99fb569fa6f045c0a8f751aced909844?sid=62e0650b-763f-431e-ae12-9f87387b1c4b

## Changes by Module

### clients/java

* **CamundaClient.java**: adds factory methods `newVariableCreationRequest()`, `newVariableUpdateRequest()`, `newVariableDeleteRequest()`.
* **New command step interfaces**: `VariableCreationRequestStep1`, `VariableUpdateRequestStep1`, `VariableDeleteRequestStep1`.
* **New response interfaces**: `CreateVariableResponse`, `UpdateVariableResponse`, `DeleteVariableResponse`.
* **Implementations**: `CreateVariableImpl`, `UpdateVariableImpl`, `DeleteVariableImpl`.
* **Response implementations**: `CreateVariableResponseImpl`, `UpdateVariableResponseImpl`, `DeleteVariableResponseImpl`.
* **Http consumers**: `DocumentDataConsumer` and `TypedApiEntityConsumer` cleaned up (simpler error handling, removed `IOException`).

### db/rdbms

* **VariableMapper.xml**: new `<delete>` statement for variables.
* **VariableMapper.java**: mapper updated accordingly.
* **VariableWriter.java**: adds `delete(...)` method wired to new mapper delete.

### zeebe/engine (processing & state)

* **EngineProcessors.java**: registers variable processors.
* **VariableBehavior.java**: extended for cluster variable semantics.
* **New processors**: `VariableCreateProcessor`, `VariableUpdateProcessor`, `VariableDeleteProcessor`.
* **Evaluation contexts**: `ClusterVariableEvaluationContext`, `CombinedEvaluationContext`, `ScopedEvaluationContext`, `VariableStateEvaluationContext`.
* **Appliers**: `VariableDeletedApplier` added, `VariableApplierV2` updated.
* **Variable state**: `MutableVariableState` extended with `createVariablePointer`, `removeVariable`, `removeVariableVariablePointer`; `DbVariableState` implements these.
* **New type**: `VariablePointer` for pointer column family.

### zeebe/exporters

* **Handlers**: new `VariableDeletedHandler`; updated `VariableHandler` to reject `DELETED` event.
* **RDBMS exporter**: `VariableExportHandler` updated to export deletions.

### zeebe/expression-language

* **EvaluationContext**: return type changed to `Object`.
* **FeelVariableContext**: adapted accordingly.

### zeebe/gateway-protocol

* **rest-api.yaml**: new endpoints

  * `POST /variables` (create)
  * `PATCH /variables/{key}` (update)
  * `DELETE /variables/{key}` (delete)

### zeebe/gateway-rest

* **VariableController.java**: new REST controller for above endpoints.
* **ResponseMapper.java**: maps new responses.

### zeebe/gateway (broker-side)

* **New broker requests**: `BrokerCreateVariableRequest`, `BrokerUpdateVariableRequest`, `BrokerDeleteVariableRequest`.

### zeebe/broker

* **CommandApiRequestReader.java**: supports `ValueType.VARIABLE` for distribution.

### zeebe/protocol-impl

* **VariableRecord.java**: default values extended for by-key operations.
* **CommandDistributionRecord.java**: supports `VARIABLE` type.

### zeebe/protocol

* **ZbColumnFamilies.java**: new column family for `VariablePointer`.
* **VariableIntent.java**: new intents `CREATE`, `UPDATE`, `DELETE` and `DELETED`.

### security/security-protocol

* **AuthorizationResourceType.java**: new resource type `CLUSTER_VARIABLE`.

### service (unified API)

* **VariableServices.java**: forwards new broker requests.
* **ApiServices.java**:  MsgPack manages single object

### qa/acceptance-tests

* **VariablesManagementTest.java**: acceptance test for create/update/delete.
* **VariableIT.java**: DB integration test for persistence & deletion.

### test-util

* **MsgPackUtil.java**: helper for single-object MsgPack payloads.

---

## Summary

* Introduces **cluster variable lifecycle management** (create, update, delete).
* Extends **client API**, **REST & broker gateway**, and **engine processors/state**.
* Adds **RDBMS exporter support** for variable deletion.
* Introduces new **evaluation contexts** for cluster-scoped variables.
* Updates **protocol** (new intents, new column family) and **security** (new resource type).
* Provides **acceptance tests** and **integration tests**.
* Cleans up some supporting utilities (`Http consumers`, `MsgPackUtil`).



## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

mathias-vandaele committed:
feat(camunda-client): Add 3 method to the camunda client to interact with the gateway to create/update/delete variables

mathias-vandaele committed:
feat(zeebe-engine): management of creation-deletion-update of variables inside the engine, creation of new column family to add pointer to variables by key, creation of all the processors, creation of new appliers, adding default values to VariableRecord, modification of the EvaluationContext, introduction of ScopedEvaluationContext, CombinedEvaluationContext and ClusterVariableEvaluationContext, management of the distribution of cluster variables

mathias-vandaele committed:
feat(db-rdbms): management of rdbms export for variable deletion

mathias-vandaele committed:
feat(security): Addtion of new resource type CLUSTER_VARIABLE for the management of cluster variable

mathias-vandaele committed:
feat(zeebe-exporter): modification of export management, addition of a new handler for deleted variable

mathias-vandaele committed:
feat(zeebe-gateway-protocol): 3 endpoint were added and all the needed resources for CREATE, UPDATE and DELETE

mathias-vandaele committed:
feat(zeebe-gateway): create all broker request object to interact with variables (CREATE, UPDATE and DELETE)

mathias-vandaele committed:
feat(zeebe-gateway-rest): add all necessary mapping inside controller to create, update and delete variables

mathias-vandaele committed:
feat(service): modify the service to send newly introduced broker request

mathias-vandaele committed:
feat(zeebe-protocol-impl): Modify the Variables record to add default values, which is necessary to be able to update and delete variable by key, addition of the VARIABLE type to the command distribution record

mathias-vandaele committed:
feat(zeebe-expression-language): modify the Evaluation context interface to return an Object, and modify the FeelVariableContext to adapt to the newly introduced change

mathias-vandaele committed:
feat(zeebe-protocol): Modify the VariableIntent to add all new Intent (DELETE, CREATE, UPDATE etc) and add a new column family to manage VariablePointer

mathias-vandaele committed:
feat(test-util): add a method so that MsgPack can manage single object

mathias-vandaele committed:
feat(qa-acceptance-test): Addition of the test for the new method in camundaClient

mathias-vandaele committed:
feat(zeebe-broker): modify CommandApiRequestReader so that ValueType VARIABLE can now be properly processed

sbuettner commented on clients/java/src/main/java/io/camunda/client/impl/command/CreateVariableImpl.java:
```diff
@@ -0,0 +1,76 @@
+
+  @Override
+  public CamundaFuture<CreateVariableResponse> send() {
+    final HttpCamundaFuture<CreateVariableResponse> result = new HttpCamundaFuture<>();
```
Should we verify that `clusterLevel` needs to be set for now and throw an error if it hasnt been set?

sbuettner commented on security/security-protocol/src/main/java/io/camunda/zeebe/protocol/record/value/AuthorizationResourceType.java:
```diff
@@ -58,6 +58,7 @@ public enum AuthorizationResourceType {
       PermissionType.DELETE_RESOURCE),
   ROLE(PermissionType.CREATE, PermissionType.READ, PermissionType.UPDATE, PermissionType.DELETE),
   SYSTEM(PermissionType.READ, PermissionType.READ_USAGE_METRIC, PermissionType.UPDATE),
+  CLUSTER_VARIABLE(PermissionType.CREATE, PermissionType.UPDATE, PermissionType.DELETE),
```
Should we also add `PermissionType.READ`?

tmetzke commented on clients/java/src/main/java/io/camunda/client/impl/CamundaClientImpl.java:
```diff
@@ -1060,6 +1066,21 @@ public VariableGetRequest newVariableGetRequest(final long variableKey) {
     return new VariableGetRequestImpl(httpClient, variableKey);
   }
 
+  @Override
+  public VariableCreationRequestStep1 newVariableCreationRequest() {
+    return new CreateVariableImpl(httpClient, jsonMapper);
+  }
+
+  @Override
+  public VariableUpdateRequestStep1 newVariableUpdateRequest() {
+    return new UpdateVariableImpl(httpClient, jsonMapper);
+  }
+
+  @Override
+  public VariableDeleteRequestStep1 newVariableDeleteRequest() {
+    return new DeleteVariableImpl(httpClient);
+  }
```
💭 For naming consistency, interactions with the engine are called `command`, while search (POST and GET by key) interactions are called `request`. This is neither written down anywhere nor enforced (because that is complex to do in the client). I believe we need to manifest this somewhere if we want this to be visible and adhered to.

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1423,21 +1423,99 @@ paths:
+      operationId: createVariable
+      summary: Create variable
+      description: |
+        Create cluster level variables
```
🔧 We're just creating one variable here.

```suggestion
        Create a cluster-level variable
```

mathias-vandaele replied:
Corrected

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1423,21 +1423,99 @@ paths:
-      operationId: getVariable
-      summary: Get variable
+      operationId: updateVariable
+      summary: Update variables
```
🔧 We're just updating one variable here.

```suggestion
      summary: Update variable
```

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1423,21 +1423,99 @@ paths:
+      operationId: deleteVariable
+      summary: Delete variable
+      description: |
+        Delete cluster level variables
```
🔧 We're just deleting one variable here.

```suggestion
        Delete a cluster-level variable.
```

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1423,21 +1423,99 @@ paths:
+      responses:
+        "201":
+          description: >
+            The variable creation status.
```
🔧 We know the result, let's describe it like this.

```suggestion
            The variable was created successfully.
```

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1423,21 +1423,99 @@ paths:
+            schema:
+              $ref: "#/components/schemas/UpdateVariableRequest"
+      responses:
+        "204":
```
❓ Do we return a result here? Otherwise, let's use 201.

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -1423,21 +1423,99 @@ paths:
+      responses:
+        "204":
+          description: >
+            The variable creation status.
```
🔧 We know the result, let's describe it like this.

```suggestion
            The variable was deleted successfully.
```

tmetzke commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -5900,6 +5978,46 @@ components:
+        propertyName: scope
+    ClusterVariableCreateRequest:
+      x-discriminator-value: cluster
+      allOf:
+        - $ref: '#/components/schemas/CreateVariableRequest'
+        - type: object
```
❓ What is the benefit if defining the request like this?

sbuettner commented on zeebe/exporters/camunda-exporter/src/main/java/io/camunda/exporter/handlers/VariableDeletedHandler.java:
```diff
@@ -0,0 +1,94 @@
+
+    entity
+        .setId(VariableForListViewEntity.getIdBy(recordValue.getScopeKey(), recordValue.getName()))
+        .setKey(record.getKey())
+        .setPartitionId(record.getPartitionId())
+        .setScopeKey(recordValue.getScopeKey())
+        .setProcessInstanceKey(recordValue.getProcessInstanceKey())
+        .setProcessDefinitionKey(recordValue.getProcessDefinitionKey())
+        .setBpmnProcessId(recordValue.getBpmnProcessId())
+        .setName(recordValue.getName())
+        .setTenantId(tenantOrDefault(recordValue.getTenantId()))
+        .setPosition(record.getPosition());
+
+    if (recordValue.getValue().length() > variableSizeThreshold) {
+      entity.setValue(recordValue.getValue().substring(0, variableSizeThreshold));
+      entity.setFullValue(recordValue.getValue());
+      entity.setIsPreview(true);
+    } else {
+      entity.setValue(recordValue.getValue());
+      entity.setFullValue(null);
+      entity.setIsPreview(false);
+    }
```
As we are only reference the entity by its `id` during deletion: 
```
batchRequest.delete(indexName, entity.getId());
```

We might not need all the entity handling here beside setting the `id`.

tmetzke reviewed (commented):
Due to time constraints, I have only looked into REST API and Java client matters here for now.

mathias-vandaele committed:
fix(code-review): fix reviewed issue from Simon buettner and Tobias Metzke-Bernstein

mathias-vandaele committed:
fix(code-review): fix reviewed issue from Simon buettner

mathias-vandaele commented on db/rdbms/src/main/resources/mapper/VariableMapper.xml:
```diff
@@ -109,6 +109,7 @@
     <constructor>
       <idArg column="VAR_KEY" javaType="java.lang.Long"/>
       <arg column="VAR_NAME" javaType="java.lang.String"/>
+      <arg column="VAR_NAME" javaType="java.lang.String"/>
```
Yes corrected

mathias-vandaele commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -5900,6 +5978,46 @@ components:
+          description: The variable name to create
+        value:
+          type: object
+          description: The variable value; string or any JSON object
```
Corrected

mathias-vandaele committed:
fix(exporter): exporter was having issues, it was including command and therefore rejection, inserting wrong entries in the secondary storage

mathias-vandaele committed:
fix(cluster-variable): add variable name validation when creating the variable

mathias-vandaele committed:
fix(license): add license header

mathias-vandaele committed:
fix(linting): fix linting error

matthewBearCamunda reviewed (commented):
The `commitlint` CI job is failing due to the commit messages not being formatted correctly. They will need to be adjusted for that CI check to pass

The commit message requirements can be found here:  https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#commit-message-guidelines

The main problems I see are related to message length and the prefix for the commit not adhereing to the style

mathias-vandaele committed:
fix(tests): fix tests on the engine and exporters

mathias-vandaele committed:
fix(tests): fix tests on the engine and exporters 2

mathias-vandaele committed:
fix(engine): implement getVariableKey

mathias-vandaele committed:
fix(engine): make test work

mathias-vandaele committed:
fix(engine): make test work

mathias-vandaele committed:
fix(engine): change zeebe-record-variable for elastic exporter

korthout committed:
fix: add read access to cluster variable resource

The readonly-admin needs read access on all resources to enable support
agent to clusters.

korthout commented on service/src/main/java/io/camunda/service/VariableServices.java:
```diff
@@ -61,4 +66,22 @@ public VariableEntity getByKey(final Long key) {
                 .getVariable(key));
   }
+
+  public CompletableFuture<VariableRecord> createClusterVariable(
+      final String name, final Object value) {
+    return sendBrokerRequest(
+        new BrokerCreateVariableRequest()
+            .clusterLevel()
+            .setVariable(name, getDocumentOrEmpty(value)));
```
❓ What happens when value is `null` or `""` (an empty string). Do we represent it as an empty document `{}`? Let's make sure we have a test that shows we can create a variable with `null` as value (end-to-end to ensure we cover both this and the engine).
